Allow RPC calls using block hash or number of latest block

cptartur opened the pull request:
## Usage related changes

<!-- How the changes from this PR affect users. -->

- Currently, in many places in the RPC, it is required that the users uses `"latest"` as block hash/number. With this PR it is also possible to use a `block_hash` or `block_number` if they belong to the latest block.
- This PR allows making calls  with block tag `"pending"` block, though it is still treated as a `"latest"` block (up to discussion if this should be included).
- Changed fixtures using  pytest.mark.parametrize usages with `indirect=True`, to parametrized fixtures.

## Development related changes

None

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/lint.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [ ] Linked the issues which this PR resolves
- [x] Updated the tests
- [ ] All tests are passing

FabijanC commented on starknet_devnet/blueprints/rpc/utils.py:
```diff
@@ -47,15 +51,30 @@ def get_block_by_block_id(block_id: BlockId) -> dict:
-            message="Calls with block_id != 'latest' are not supported currently.",
-        )
+    last_block = state.starknet_wrapper.blocks.get_last_block()
+    print(type(block_id), block_id)
```
print?

FabijanC commented on starknet_devnet/blueprints/rpc/utils.py:
```diff
@@ -47,15 +51,30 @@ def get_block_by_block_id(block_id: BlockId) -> dict:
-            code=-1,
-            message="Calls with block_id != 'latest' are not supported currently.",
-        )
+    last_block = state.starknet_wrapper.blocks.get_last_block()
```
This last block retrieval can be moved inside of the `if isinstance(block_id, dict)` clause

FabijanC commented on starknet_devnet/blueprints/rpc/utils.py:
```diff
@@ -47,15 +51,30 @@ def get_block_by_block_id(block_id: BlockId) -> dict:
+
+    raise RpcError(
+        code=-1,
+        message="Calls with block_id != 'latest' are not supported currently.",
```
The error message should also be modified, right? If I may suggest extracting the valid block_ids to a variable and using it for both the actual checking and the message construction.

FabijanC reviewed (changes requested):
Thanks for the contribution!

1. Accepted block IDs can be documented in `page/docs/guide/json-rpc-api.md`.
2. Currently only `"latest"` is tested. Shouldn't providing `"pending"` also be tested?

cptartur commented on starknet_devnet/blueprints/rpc/utils.py:
```diff
@@ -47,15 +51,30 @@ def get_block_by_block_id(block_id: BlockId) -> dict:
+    print(type(block_id), block_id)
+
+    if isinstance(block_id, dict):
+        if "block_hash" in block_id:
```
I think it's not handled anywhere, I'll include a test for this and a proper check.

cptartur commented:
1. Will be added
2. I wasn't sure if we want to include "pending", for me it seems like a logical addition since `"pending"` is supported in gateway as far as I know. I'll extend the test to cover pending block.

cptartur commented:
Addressed the comments. I've also unchecked "tests are passing" because `test/test_postman.py::test_postman` suddenly stopped working. I suspect it's a problem with some local configuration on my end so if tests could be run on your end we could verify that.

FabijanC commented on page/docs/guide/json-rpc-api.md:
````diff
@@ -31,3 +31,31 @@ Response:
+
+Methods that require a `block_id` only support ids of the `latest` or `pending` block:
+
+```json
````
`json` doesn't support comments, so if you could change the language to `js` or `ts`.

FabijanC commented on page/docs/guide/json-rpc-api.md:
```diff
@@ -31,3 +31,31 @@ Response:
 
 - `starknet_protocolVersion` - will be removed in a future version of the specification
+
+Methods that require a `block_id` only support ids of the `latest` or `pending` block:
```
Can you point out that these tags refer to the same block?

FabijanC commented on test/rpc/test_rpc_call.py:
```diff
@@ -52,6 +52,29 @@ def test_call_raises_on_incorrect_contract_address():
+@pytest.mark.usefixtures("run_devnet_in_background", "deploy_info")
+def test_call_raises_on_both_hash_and_number():
+    """
+    Call contract with incorrect address
```
Docstring not corresponding to test.

FabijanC reviewed (changes requested):
Cloned your fork, postman test is passing.
